### PR TITLE
(feat) Introduce locale/Openmrs datepicker

### PR DESCRIPTION
**Pull request by kajambiya**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
This PR introduces the `OpenmrsDatepicker` component from openmcore/esm-styleguide. Since the Ethiopian team already uses this component in patient registration, there's no need for them to register a different datepicker in Form Engine.

## Screenshots

https://github.com/openmrs/openmrs-form-engine-lib/assets/12844964/017df011-ca0e-4d4e-967e-27a304687c51

### CODE REVIEW

**Comment by arodidev** [comment]
Thanks @kajambiya, mind resolving the conficts and build failure?

File: src/form-engine.test.tsx [hunk 1]
```diff
@@ -649,7 +649,7 @@ describe('Form engine component', () => {
       });
     });
 
-    it('should only show question when age is under 5', async () => {
+    fit('should only show question when age is under 5', async () => {
```
Comment (samuelmale) [inline src/form-engine.test.tsx]: Ditto

File: package.json [hunk 2]
```diff
@@ -52,7 +52,7 @@
   },
   "devDependencies": {
     "@carbon/react": "^1.37.0",
-    "@openmrs/esm-framework": "next",
+    "@openmrs/esm-framework": "^5.6.1-pre.1880",
```
Comment (samuelmale) [inline package.json:55]: Can you revert this change?

File: package.json [hunk 3]
```diff
@@ -86,7 +86,7 @@
     "jest-environment-jsdom": "^29.7.0",
     "jest-when": "^3.6.0",
     "lint-staged": "^15.2.2",
-    "openmrs": "next",
+    "openmrs": "^5.6.1-pre.1880",
```
Comment (samuelmale) [inline package.json:89]: Ditto

File: src/components/inputs/date/date.component.tsx [hunk 4]
```diff
@@ -121,36 +123,30 @@ const DateField: React.FC<FormFieldProps> = ({ question, onChange, handler, prev
...
+                value={field.value}
+                carbonOptions={{
+                  placeholder: placeholder,
+                  warn: warnings[0]?.message,
+                  warnText: warnings[0]?.message,
+                  className: styles.boldedLabel,
+                  datePickerType: datePickerType,
+                }}
+                {...(typeof question.readonly === 'boolean' && { readonly: question.readonly })}
+                {...(typeof question.disabled === 'boolean' && { disabled: question.disabled })}
```
Comment (samuelmale) [inline src/components/inputs/date/date.component.tsx:148]: Would this work?

```suggestion
                disabled={question.isDisabled}
                readonly={question. readonly}
```

Comment (kajambiya) [reply src/components/inputs/date/date.component.tsx:148]: Not really
In RFE, the type of **question.readonly** is `string | boolean` and that of **question.disabled** is `boolean | Omit<DisableProps, "isDisabled">` whereas in core **readonly** and **disabled** are both of type `boolean.`
The solution was either to allow the `OpenmrsDatepicker props` in core accept the same exact types or handle them at FE level. 
I tried to do it in core but according to [this](https://github.com/openmrs/openmrs-esm-core/pull/1007#discussion_r1612280735) comment from @ibacher, this should be handled in FE and core only receives boolean types.

When I went through the code, I realized that for `readonly` for example, it accepts a string (expression) but after the expression is evaluated, the value is changed to a boolean. Given this background, it made sense to only pass these 2 if the data type is boolean so that the types in core stay intact.

File: package.json (hunk 3 above)
Comment (kajambiya) [reply package.json]: Screenshot 2024-06-06 at 13 07 11
When I change these back to next, the CI fails(see screenshot)

File: src/components/inputs/date/date.component.tsx (hunk 4 above)
Comment (samuelmale) [reply src/components/inputs/date/date.component.tsx:148]: > In RFE, the type of question.readonly is string | boolean and that of question.disabled is boolean | Omit<DisableProps, "isDisabled"> whereas in core readonly and disabled are both of type boolean.

For the disabled property you're supposed to use `question.isDisabled` which is type boolean.

As for `readonly` and other related props we should update the typing to simply `boolean` because the [schema transformer parses the strings to booleans](https://github.com/openmrs/openmrs-form-engine-lib/blob/main/src/transformers/default-schema-transformer.ts#L79).

Comment (kajambiya) [reply src/components/inputs/date/date.component.tsx:148]: Just checked the schema `parseBooleanTokenIfPresent` function. 
From my understanding, it assumes that the the string is either "true" or "false"
However, we have cases where it is a totally different expression (I believe this wasn't a smart implementation) and I've seen it used in PMTCT forms.
Here is [how FE handles it](https://github.com/openmrs/openmrs-form-engine-lib/blob/main/src/components/encounter/encounter-form.component.tsx#L309C11-L322C12)
We need to agree if this is no longer desirable so that the forms where it is being used can also be adjusted accordingly.

Comment (samuelmale) [reply src/components/inputs/date/date.component.tsx:148]: You make a good point but at this point it's still fair to assume that the `readonly` prop has already been processed to a boolean. As you can see in the snippet below the expression is stored in `field['readonlyExpression']`. (I think to be more consistent we should use the `readonly` vs `isReadonly` pattern)

```ts
if (typeof field.readonly == 'string' && field.readonly?.split(' ')?.length > 1) {
            // needed to store the expression for further evaluations
            field['readonlyExpression'] = field.readonly;
            field.readonly = evaluateExpression(

```

To make the TS compiler happy, use the `isTrue(string | boolean)` function:

```ts
readonly={isTrue(question.readonly)}
```

Comment (kajambiya) [reply src/components/inputs/date/date.component.tsx]: This makes sense.

Done.

**Review by samuelmale — APPROVED** [review]
LGTM
